Actually enable typechecking for _inductor/index_propagation.py

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110112
* __->__ #110110

It was supposed to be enabled in #105622 but that PR neglected to update
.lintrunner.toml.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
Actually enable typechecking for _inductor/index_propagation.py

It was supposed to be enabled in #105622 but that PR neglected to update
.lintrunner.toml.

[ghstack-poisoned]

eellison commented on torch/_inductor/index_propagation.py:
```diff
@@ -176,22 +176,32 @@ def unwrap(self, a: Union[Any, IndexPropVar]) -> Any:
 
-    def wrap(self, a: Any) -> Union[IndexPropVar, Sequence[IndexPropVar]]:
+    # FIXME MYPYNOFOLLOW thinks this overload clashes with the one below because it doesn't follow the import of
+    # sympy, so it treats sympy.Symbol as an Any type. We can probably remove the type-ignore once we remove the nofollow.
```
does importing `Symbol` from `sympy` help with this ? alternatively, is enabling MYPY FOLLOW for this file easier ?

int3 replied:
> does importing Symbol from sympy help with this ?

Nope, `follow_imports=skip` in mypy-nofollow.ini ensures that no imports are ever followed.

> alternatively, is enabling MYPY FOLLOW for this file easier ?

Lintrunner seems to only support globbing files via include-then-exclude, i.e. the set difference of `include_patterns - exclude_patterns`. The regular MYPY config includes all torch/ files, then excludes all torch/_inductor files. So if I wanted to include just one `_inductor/` file I would have to explicitly exclude all the other inductor files that don't typecheck in follow-imports mode, which is kind of awkward.

I could set up a separate lintrunner section just for this file, but that seems like overkill.

Setting `follow_imports=normal` in mypy-nofollow.ini makes this file typecheck cleanly, but unsurprisingly causes a deluge of errors in the other inductor files. Side note: A lot of those typing errors look legit, i.e. at least mis-typing if not an actual runtime error. We should really drop the 'nofollow' thing as soon as possible...

int3 replied:
That said, all this `@overload`-ing avoids just one type error in the body of `wrap`. It would be a lot more succinct to suppress that one error via a type-ignore comment instead of specifying these overloads, which in any case require their own type-ignore comment, plus a comment block explaining why. We could always add the overloads back when we enable import following.

I'm actually leaning towards this option now. Wdyt?

int3 replied:
Scratch all of the above, I've redone the typing to a form that's hopefully better and requires no type-ignore comments at all. I've added a recursive type def `IndexPropExpr` that covers all the possible return values of `wrap`. I'm not sure `IndexPropExpr` is the best name for it though... do you or @peterbell10 have any thoughts?

peterbell10 replied:
It's not really an expression, maybe `IndexPropResult` would make more sense?

eellison reviewed (commented):
have one question

int3 committed:
Update on "Actually enable typechecking for _inductor/index_propagation.py"


It was supposed to be enabled in #105622 but that PR neglected to update
.lintrunner.toml.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 committed:
Update on "Actually enable typechecking for _inductor/index_propagation.py"


It was supposed to be enabled in #105622 but that PR neglected to update
.lintrunner.toml.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 committed:
Update on "Actually enable typechecking for _inductor/index_propagation.py"


It was supposed to be enabled in #105622 but that PR neglected to update
.lintrunner.toml.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 committed:
Update on "Actually enable typechecking for _inductor/index_propagation.py"


It was supposed to be enabled in #105622 but that PR neglected to update
.lintrunner.toml.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/index_propagation.py:
```diff
@@ -144,6 +144,9 @@ def __post_init__(self):
         ), "Symbolic IndexPropVar must contain a TypedExpr"
 
 
+IndexPropResult = Union[IndexPropVar, Tuple["IndexPropResult", ...]]
```
```suggestion
IndexPropResult : TypeAlias = Union[IndexPropVar, Tuple["IndexPropResult", ...]]
```
import TypeAlias from typing_extensions

int3 committed:
Update on "Actually enable typechecking for _inductor/index_propagation.py"


It was supposed to be enabled in #105622 but that PR neglected to update
.lintrunner.toml.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 commented:
@pytorchbot merge
